### PR TITLE
Query ergonomics

### PR DESCRIPTION
**Pull request by gciuloaica**
Implementation that supports:
- up to 22 parameters in Case Class
- unlimited cases in a sum type - but provides support only for `case object`
- primitives could be wrapped in `Option` or `List`

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/QueryParameters.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package zhttp.http
+
+final case class QueryParameters(raw: Map[String, List[String]]) { self =>
```
Comment (smehta91) [inline zio-http/src/main/scala/zhttp/http/QueryParameters.scala:3]: This should use `DynamicValue` of ZSchema

File: zio-http/src/main/scala/zhttp/http/QueryParameters.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def ++(other: QueryParameters): QueryParameters = self.combine(other)
+
+  def combine(other: QueryParameters): QueryParameters = QueryParameters(self.raw ++ other.raw)
+
+  def queryParams: QueryParameters = self
+}
```
Comment (smehta91) [inline zio-http/src/main/scala/zhttp/http/QueryParameters.scala:9]:
```suggestion
final case class QueryParameters(raw: ListMap[String, DynamicValue]) { self =>
  def ++(other: QueryParameters): QueryParameters = self.combine(other)

  def combine(other: QueryParameters): QueryParameters = QueryParameters(self.raw ++ other.raw)

  def queryParams: QueryParameters = self
}

object QueryParameters {

  val empty: QueryParameters = QueryParameters(ListMap.empty[String, DynamicValue])

}
```

File: zio-http-query-params/src/main/scala/zhttp/http/query/QueryParams.scala [hunk 3]
```diff
@@ -0,0 +1,1669 @@
...
+
+import java.time._
+import java.util.UUID
+import scala.util.Try
+
+object QueryParams {
+
+  implicit class QueryParametersWrapper(queryParams: QueryParameters) {
+    def decode[A](implicit schema: Schema[A]): Either[String, A] =
+      CaseClassDecoder.decode(schema, queryParams.raw)
```
Comment (smehta91) [inline zio-http-query-params/src/main/scala/zhttp/http/query/QueryParams.scala:15]: `schema.fromDynamic(DynamicValue.Record(queryParams.raw))`

**Review by smehta91 — CHANGES_REQUESTED** [review]
We can build `DynamicValue` using QueryParameter domain and while decoding we can decode typed value if we have schema for the type

**Comment by gciuloaica** [comment]
new API in progress https://github.com/dream11/zio-http/pull/1209

**Pull request closed without merging**